### ISSUE TITLE
Feature request: Player-owned fighters could bought back as part of a salvaging mechanic

### ISSUE THREAD

**Issue by samrocketman** [endless-sky/endless-sky#6861]
### Problem Description

Discord discussion: https://discordapp.com/channels/251118043411775489/947227869036286033/978644890189701180

Feature: Player ability to purchase destroyed player-owned fighters on landing.  I think it would be cool of the player got a "salvager's" dialog where they can buy back lost player-owned ships (specifically fighters and drones).

The intent of this feature is to limit the scope (initially) only to player-owned fighters and drones.

### Related Issue Links

- https://github.com/endless-sky/endless-sky/issues/4034
- ([link to comment][1]) @Tadrix mentions that regenerating fighters would be only a good idea for automata since fighters have pilots.  This would not be an issue if the ship was re-purchased on a planet where player ships replenish crew.
- ([link to comment][2]) @MessyMix mentions that drones should be outfits instead of actual ships.  More like weapons with ammo than actual ships that can be customized.  I like being able to customize fighters so this idea isn't attractive to me.
  - ([link to comment][3]) @Tadrix also says the main issue is how tedious it is for a player to customize fighters and to offers to repurchase the exact configuration.  This issue spawned from this idea.  To re-buy player-owned fighters which were destroyed at their exact configuration.
  - ([link to comment][4]) @petervdmeer expands on @Tadrix idea and gives a hint at implementation that player ships "pruning" could be skipped so that the player could buy their exact ship back outfits and all.  This is how I intend to implement this feature.

The common theme across all of this is ship customization is fun.  Losing customized fighters is not fun.  Players should be able to recover customized fighters but how could it be done in a way that isn't too immersion breaking.  I think salvaging is one way this could be done and there's some comments on implementation.

[1]: https://github.com/endless-sky/endless-sky/issues/4034#issuecomment-441511848
[2]: https://github.com/endless-sky/endless-sky/issues/4034#issuecomment-441511848
[3]: https://github.com/endless-sky/endless-sky/issues/4034#issuecomment-442085239

### Desired Solution

Upon landing a player gets a pop up dialog similar to refilling ammo in order to be given the option to re-buy their own destroyed fighters (by a hypothetical salvaging NPC that the player never sees in-game).

### Landing Dialog

The dialog would look something like:

> Some salvagers picked up your lost drones and fighters.  Would you like to buy them back for X credits?

Upon which, paying the fee means the player effectively bought back their own ships that they customized.

### Uniques become useful in battle

This has the benefit of using the Emerald Sword's fighters which have 100% unique outfits.  The player can play with them without fear of loss as long as they can buy them back upon landing.  Choosing not to buy them or not affording them means the ships should be deleted normally.

### Preference

This behavior could be locked behind a preference with the default being "off" (the way it is today in-game).  The preference could be named: `Destroyed fighters can be salvaged` on/off which gives the option for the player to buy them back if they can afford it.

### Universality

This feature, to keep gameplay fun, should be universal across all governments and inhabited planets.  Similar to the player being able to buy outfits with the "universal credits".  It's magically universal but necessary to keep implementation simple and playability fun.

### Scope

Initially, this should only be scoped to fighters and drones which are player owned.  There was some discussion on including all ships but I want the scope limited to fighters and drones for this initial implementation to keep it simple, small-ish in scope.

Future-looking I would like to expand on the salvaging as a general game mechanic but let's limit this for now.

### Alternative Approaches

There's several alternatives discussed in the issue I referenced but there's several issues pointed out.

I would like to open a separate feature request for alien automata manufacturing their own fighters/drones as a feature of a future yet to be determined alien race for content creators.

**Comment by samrocketman** [comment]
Once implementation is settled I plan to implement this.

**Comment by samrocketman** [comment]
@petervdmeer can you expand on this?  So that I know better how it could be implemented.

Also, would it be a problem if I made the preference have the option to be 100% recovery if the player can afford it.

So something like frugal ammo preference:
- No recovery
- Chance of recovery
- Always recover

Where "always" means the player can buy it back if they can afford it.  Future looking I could see a salvage panel in shipyards and outfitters where a player can buy from salvagers instead if outfitters for some things.

But since I want to initially avoid a lot of UI work I thought of this popup similar to ammo.

### Some background

I find myself reloading the game regardless if progress (30min mining or capturing or battle doesn't matter) if I lose even one fighter.  Because customizing fighters (and re-buying fighters to customize again) is so expensive time-wise that I opt to reload the game all of the time.

It would also mean I could use unique fighters like emerald sword fighters which until now I just leave parked.

**Comment by daveeater** [comment]
Should it be somehow range-limited, as in, a player can buy back a lost fighter in the system it was lost, or in a nearby system with a spaceport, but the player cannot buy back the salvaged fighter somewhere halfway across the galaxy. It only makes sense to me if the salvager is based near the point of fighter loss - no salvager is going to drag a salvaged fighter from the Rim to a Paradise planet just because the player chose to go there before landing.

**Comment by petervdmeer** [comment]
> @petervdmeer can you expand on this? So that I know better how it could be implemented.

I was thinking of something like this:
```
ship "Recoverable Sparrow"
	recoverable 1

ship "Somewhat Recoverable Sparrow"
	recoverable .5

ship "Non-Recoverable Sparrow"
	recoverable 0
```

**Comment by samrocketman** [comment]
@petervdmeer that makes sense thanks

**Comment by petervdmeer** [comment]
I'm also not sure about the "salvagers" dialog; just showing the "lost" ships for sale again in the shipyard seems more neutral.

Another direction we could take is to make this a more content-driven feature, similar to the give-ship command, but then for "giving back" one or more ships that the player lost. (Or for making lost-ships available at the shipyard again.) Then we can have the content-writers write all the stories on why lost ships suddenly become available.

And should lost ships always be given back at the planet where the player happens to be? Or would we also want to have the option to give such ships back near the place where the ship was lost? (Suggesting that the player needs to travel there to pickup the ships.)

Would the reason for loss also matter? Captured ships would/could be given back in a different way than ships that were destroyed.

**Comment by samrocketman** [comment]
> Should it be somehow range-limited

It could be it isn't implemented, yet.

> I'm also not sure about the "salvagers" dialog; just showing the "lost" ships for sale again in the shipyard seems more neutral.

The way shipyards are implemented ships are either for sale or not from what I can tell.  This limitation would have to be worked around to be more dynamic i.e. track which ships were lost and make sure the player couldn't buy more than they lost.  I feel like this would be much more complex than essentially giving back what the player already had for a fee.

**Comment by samrocketman** [comment]
> no salvager is going to drag a salvaged fighter from the Rim to a Paradise planet just because the player chose to go there before landing.

I agree with this.  At the same time though I think some realism can be sacrificed for good gameplay.  An example I mentioned was the universal credits. Another example is "ship speed limits" and ship "drag" in space.  Realism doesn't always make for good gameplay which is why I am proposing this as a generic universal mechanic to improve the impact of losing customized fighters.  I feel like paying a fee to get fighters back is much better than the grind of losing real life time which is why people avoid fighters in general.

**Comment by DeciduousBeard** [comment]
If this is independent of shipyards, so can be offered on any inhabited planet, the dialog could state that the salvagers have quickly repaired your equipment in their own workshop and are offering it to you as the original owner for a one-off discounted price. This means it can be implemented as a simple mission on landing with a yes/no. If multiple classes of ships were lost, multiple missions could pop up in sequence "Salvagers have found 3 fighters they believe belonged to you", then "Salvagers rescued your Argosy". There should be no need to travel to the nearest shipyard.

The "lore" behind this could be that salvagers historically were often viewed as thieves, which was not helped by the fact that many worked as both pirates and salvagers. So in recent years salvagers have adopted a policy of seeking out the original owner of all salvaged equipment, and offering it back to them first at a one-off discounted price. This makes good business sense because a quick sale is likely if you can find the original owner and sell it back to them, and a quick sale brings in immediate guaranteed revenue so is worth offering a discount to achieve. If the owner does not purchase the ship they are sold wholesale to shipyards or broken up and the parts sold piecemeal to outfitters, so essentially disappear at that point into the general economy - explaining very reasonably why this is a one-off offer and then disappears never to be seen again.

> And should lost ships always be given back at the planet where the player happens to be? Or would we also want to have the option to give such ships back near the place where the ship was lost? (Suggesting that the player needs to travel there to pickup the ships.)

Knowing that their best chance of a quick sale is to find the original owner, it is entirely reasonable for salvagers to follow you for up to (say) 3 jumps and then tap you on the shoulder at the next planet you land on offering to sell you back the ships they have salvaged. So this can be offered at the next planet landed, any type of planet, within a set distance of the system the ship was lost in.

> Also, would it be a problem if I made the preference have the option to be 100% recovery if the player can afford it.
> 
> So something like frugal ammo preference:
> 
> * No recovery
> * Chance of recovery
> * Always recover

Might I suggest that there is a hidden probability for this, not everything will be salvaged, but the likelihood of a small ship being salvaged is greater than a large ship (logic being a small ship can be dragged into the bay of a large salvage ship easily, but salvaging a large ship is a lot of work). For instance, fighters have a 90% probability of being salvaged, interceptors 60%, and there's something like a 10% chance of something like an Argosy or larger being salvaged. That would mean that very occasionally a large ship the player lost would be salvaged, and this would be an unusual event that would justify an interesting dialog explaining the circumstances. Keeping this as an interesting thing that pops up occasionally, and not clutter that keeps filling the screen every time you land after a battle.

**Comment by DeciduousBeard** [comment]
> we introduce this mechanic, then I would like it to be a generic configuration item in the ships definitions. So we can set per ship-type which ships can be bought back by this mechanism and which ones not. (maybe even with a "recover factor" that gives the chance of a destroyed ship getting offered for sale back after salvage).

Would it not be much simpler to just say that all sparrows have a certain probability of recovery? Making new ship types for this would be far harder to integrate with existing missions.

> * @Tadrix mentions that regenerating fighters would be only a good idea for automata since fighters have pilots.  This would not be an issue if the ship was re-purchased on a planet where player ships replenish crew.

There is no need for this to occur only on planets where crew are available to hire. The salvage team itself would contain many persons, enough to pilot the ships they have rescued (how did they fly them to you to sell them?), some of whom may be quite happy to sign on with you as crew. I can imagine an exciting career path for a young person on a poor planet being to work for a salvage company repairing and learning to pilot captured ships, in the hope of being hired by a captain the ship was sold to and getting to explore the galaxy. Meaning they'd have very high staff turnover. So you can just offer the ships for sale on any planet, and if an explanation is needed say that there are some members of the salvage team who would love to sign on with you as crew and see the galaxy.

**Comment by petervdmeer** [comment]
> Would it not be much simpler to just say that all sparrows have a certain probability of recovery? Making new ship types for this would be far harder to integrate with existing missions.

Yes, that is what I wanted to express. Sorry for my bad example, I didn't want to suggest defining multiple ship types.

**Comment by samrocketman** [comment]
@petervdmeer should salvage rely on a preference or be solely up to the data files and content creators?

Either way the data files approach works for me because I can write a plugin to make ships salvageable

**Comment by samrocketman** [comment]
@DeciduousBeard an interesting review; thanks!

Now some questions (for everyone responding).

Regarding @DeciduousBeard 's statement with regard to pop-ups.

> and not clutter that keeps filling the screen every time you land after a battle.

And @petervdmeer 's statement on probability for salvage...

- If the random probability for salvage is not met for a destroyed ship, then what should happen to it?  Should it be assumed not salvagable and deleted from memory?
- How often should the probability roll for a ship?  Every time a player lands?
- Or should salvage probability only be calculated once and the dialog pop-up be random?
  - I'm thinking once upon ship destruction to set a variable like `canBeSalvaged` along side a variable `salvageOffered`.  If `salvageOffered == true and IsDestroyed()` then the ship can be deleted from memory.

**Comment by petervdmeer** [comment]
> @petervdmeer should salvage rely on a preference or be solely up to the data files and content creators?

Leave it solely up to the data files and content creators sounds much better to me.

**Comment by samrocketman** [comment]
@petervdmeer per your data file suggestion should the default chance of salvage be 0% since that is vanilla today?

@DeciduousBeard 
You mentioned
> Knowing that their best chance of a quick sale is to find the original owner, it is entirely reasonable for salvagers to follow you for up to (say) 3 jumps and then tap you on the shoulder at the next planet you land on offering to sell you back the ships they have salvaged.

So when I mentioned random dialog popup I thought you meant a ship salvaging could occur for the player landing up to 3 times.

---

Should there be a `salvage jump radius` as well?

**Comment by samrocketman** [comment]
If there should be a jump radius should a min/max radius be supported?

**Comment by DeciduousBeard** [comment]
> So when I mentioned random dialog popup I thought you meant a ship salvaging could occur for the player landing up to 3 times.

What I meant was that it should happen the next time the player landed, if they land in any system up to a reasonable radius (e.g. three jumps), but not if they travel further than that before landing because:

> no salvager is going to drag a salvaged fighter from the Rim to a Paradise planet just because the player chose to go there before landing.

**Comment by Saturnalia9** [comment]
> There is no need for this to occur only on planets where crew are available to hire. The salvage team itself would contain many persons, enough to pilot the ships they have rescued (how did they fly them to you to sell them?), some of whom may be quite happy to sign on with you as crew. I can imagine an exciting career path for a young person on a poor planet being to work for a salvage company repairing and learning to pilot salvaged ships, in the hope of being hired by a captain the ship was sold to and getting to explore the galaxy. Meaning they'd have very high staff turnover. So you can just offer the ships for sale on any planet, and if an explanation is needed say that there are some members of the salvage team who would love to sign on with you as crew and see the galaxy.

That would be an interesting dialog to see. Like if someone on a pirate world wanted to pilot one of my ships so they could escape the piracy and poverty? I'd 100% hire them.
